graylog2-server/src/main/java/org/graylog2/migrations/V20191121145100_FixDefaultGrokPatterns.java

```diff
@@ -0,0 +1,150 @@
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.auto.value.AutoValue;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import org.graylog.autovalue.WithBeanGetter;
+import org.graylog2.grok.GrokPattern;
+import org.graylog2.grok.GrokPatternService;
+import org.graylog2.plugin.cluster.ClusterConfigService;
+import org.graylog2.plugin.database.ValidationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Fixes errors in the default grok patterns that have been installed by the
+ * {@link V20180924111644_AddDefaultGrokPatterns} migration.
+ * <p>
+ * At the moment this is only the COMMONAPACHELOG pattern, but if we find more issues with the default patterns we
+ * <em>should</em> be able to adjust this migration by adding more patterns instead of having to write additional
+ * migrations.
+ *
+ */
+public class V20191121145100_FixDefaultGrokPatterns extends Migration {
+    private static final Logger log = LoggerFactory.getLogger(V20191121145100_FixDefaultGrokPatterns.class);
+
+    private final ClusterConfigService configService;
+    private final GrokPatternService grokPatternService;
+
+    @VisibleForTesting
+    static final List<PatternToMigrate> patternsToMigrate = ImmutableList.of(
+        PatternToMigrate.builder()
+                      .name("COMMONAPACHELOG")
+                      .migrateFrom("%{IPORHOST:clientip} %{HTTPDUSER:ident} %{USER:auth} \\[%{HTTPDATE:timestamp}\\] \"(?:%{WORD:verb} %{NOTSPACE:request}(?: HTTP/%{NUMBER:httpversion})?|%{DATA:rawrequest})\" %{NUMBER:response} (?:%{NUMBER:bytes}|-)")
+                      .migrateTo("%{IPORHOST:clientip} %{HTTPDUSER:ident} %{USER:auth} \\[%{HTTPDATE:timestamp;date;dd/MMM/yyyy:HH:mm:ss Z}\\] \"(?:%{WORD:verb} %{NOTSPACE:request}(?: HTTP/%{NUMBER:httpversion})?|%{DATA:rawrequest})\" %{NUMBER:response} (?:%{NUMBER:bytes}|-)")
+                      .build()
+    );
+
+    @Inject
+    public V20191121145100_FixDefaultGrokPatterns(final ClusterConfigService clusterConfigService,
+                                                  GrokPatternService grokPatternService) {
+        this.grokPatternService = grokPatternService;
+        this.configService = clusterConfigService;
+    }
+
+    @Override
+    public ZonedDateTime createdAt() {
+        return ZonedDateTime.parse("2019-11-21T14:51:00Z");
+    }
+
+    @Override
+    public void upgrade() {
+        final MigrationCompleted migrationCompleted = configService.get(MigrationCompleted.class);
+
+        final Set<String> patternNames = patternsToMigrate.stream()
+                .map(PatternToMigrate::name)
```

Comment:
One small nitpick: If we want to write a second migration for the same pattern, we cannot use the name here, because it would be skipped. But a second identifier could still be introduced when need it :)